shuttle/src/runtime/failure.rs

```diff
@@ -4,76 +4,54 @@
-//! 1. If a panic occurs within `ExecutionState`, the panic hook might not be able to access the
-//!    execution state to retrieve the failing schedule, so we need to be careful about accessing it
-//!    and try to recover from this problem when the panic is later caught in
-//!    `ExecutionState::step`. We try wherever possible to avoid panicking in this state, but if it
-//!    does happen we want to get useful output and not crash.
-//! 2. In addition to simply printing the failing schedule, we want to include it in the panic
-//!    payload wherever possible, so that we can parse it back out in tests (essentially we are
-//!    reinventing try/catch, which is usually an anti-pattern in Rust, but here we don't have
-//!    control over the user code that panics). That means sometimes we end up calling
-//!    `persist_schedule` twice -- once in the panic hook (where we can't modify the panic payload)
-//!    and again when we catch the panic (where we can modify the payload). We don't want to print
-//!    the schedule twice, so we keep track of whether the info has already been printed.
-
+use std::cell::Cell;
 use std::fs::OpenOptions;
 use std::io::{ErrorKind, Write};
 use std::panic;
 use std::path::{Path, PathBuf};
-use std::sync::{Mutex, Once};
+use std::sync::Once;
 
-use crate::runtime::execution::ExecutionState;
+use crate::runtime::execution::{CurrentSchedule, ExecutionState};
 use crate::scheduler::serialization::serialize_schedule;
-use crate::scheduler::Schedule;
 use crate::{Config, FailurePersistence};
 
-/// Produce a message describing how to replay a failing schedule.
-///
-/// If `print_if_fresh` is true, the message will also be printed to stderr if this is the first
-/// time `persist_failure` has been called.
-pub fn persist_failure(schedule: &Schedule, message: String, config: &Config, print_if_fresh: bool) -> String {
-    // Disarm the panic hook so that we don't print the failure twice
-    if let PanicHookState::Persisted(persisted_message) = PANIC_HOOK.with(|lock| lock.lock().unwrap().clone()) {
-        return persisted_message;
-    }
-
-    let persisted_message = persist_failure_inner(schedule, message, config);
-    PANIC_HOOK.with(|lock| *lock.lock().unwrap() = PanicHookState::Persisted(persisted_message.clone()));
-    if print_if_fresh {
-        eprintln!("{persisted_message}");
-    }
-    persisted_message
-}
-
-/// Produce a message describing how to replay a failing schedule that ended on a given task.
-pub fn persist_task_failure(schedule: &Schedule, task_name: String, config: &Config, print_if_fresh: bool) -> String {
-    persist_failure(
-        schedule,
-        format!("test panicked in task '{task_name}'"),
-        config,
-        print_if_fresh,
-    )
+// When we last persisted a schedule. Used so that we don't persist the same schedule twice.
+thread_local! {
+    static SCHEDULE_PERSITED_AT: Cell<usize> = const { Cell::new(0) };
```

Comment:
typo: "PERSISTED" misspelled

Also: why store the length and not just a boolean?

Reply:
So that we don't persist the same schedule twice. We want to persist longer schedules. An issue now is that we persist the schedule, then keep scheduling, and never persist the longer schedule. The second issue now is that a caught panic will trigger a schedule serialization but the actual test panic which comes later will go unserialized.